audit/hashstructure.go

```diff
@@ -19,107 +19,169 @@ func HashString(salter *salt.Salt, data string) string {
-// it will be passed through.
-//
-// The structure is modified in-place.
-func Hash(salter *salt.Salt, raw interface{}, nonHMACDataKeys []string) error {
+// Hash will copy and hash the given type. This has built-in support for auth,
+// requests, responses, and ResponseWrapInfo. If it is given a type that isn't
+// recognized, then it will be passed through unchanged.
+func HashAuth(salter *salt.Salt, in *logical.Auth, HMACAccessor bool) (*logical.Auth, error) {
+	if in == nil {
+		return nil, nil
+	}
+
 	fn := salter.GetIdentifiedHMAC
+	auth := *in
 
-	switch s := raw.(type) {
-	case *logical.Auth:
-		if s == nil {
-			return nil
-		}
-		if s.ClientToken != "" {
-			s.ClientToken = fn(s.ClientToken)
-		}
-		if s.Accessor != "" {
-			s.Accessor = fn(s.Accessor)
-		}
+	if auth.ClientToken != "" {
+		auth.ClientToken = fn(auth.ClientToken)
+	}
+	if HMACAccessor && auth.Accessor != "" {
+		auth.Accessor = fn(auth.Accessor)
+	}
+	return &auth, nil
+}
 
-	case *logical.Request:
-		if s == nil {
-			return nil
-		}
-		if s.Auth != nil {
-			if err := Hash(salter, s.Auth, nil); err != nil {
-				return err
-			}
-		}
+func HashRequest(salter *salt.Salt, in *logical.Request, HMACAccessor bool, nonHMACDataKeys []string) (*logical.Request, error) {
+	if in == nil {
+		return nil, nil
+	}
 
-		if s.ClientToken != "" {
-			s.ClientToken = fn(s.ClientToken)
-		}
+	fn := salter.GetIdentifiedHMAC
+	req := *in
 
-		if s.ClientTokenAccessor != "" {
-			s.ClientTokenAccessor = fn(s.ClientTokenAccessor)
+	if req.Auth != nil {
+		cp, err := copystructure.Copy(req.Auth)
+		if err != nil {
+			return nil, err
 		}
 
-		data, err := HashStructure(s.Data, fn, nonHMACDataKeys)
+		req.Auth, err = HashAuth(salter, cp.(*logical.Auth), HMACAccessor)
 		if err != nil {
-			return err
+			return nil, err
 		}
+	}
 
-		s.Data = data.(map[string]interface{})
+	if req.ClientToken != "" {
+		req.ClientToken = fn(req.ClientToken)
+	}
+	if HMACAccessor && req.ClientTokenAccessor != "" {
+		req.ClientTokenAccessor = fn(req.ClientTokenAccessor)
+	}
 
-	case *logical.Response:
-		if s == nil {
-			return nil
-		}
+	data, err := hashMap(fn, req.Data, nonHMACDataKeys)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Data = data
+	return &req, nil
+}
 
-		if s.Auth != nil {
-			if err := Hash(salter, s.Auth, nil); err != nil {
-				return err
+func hashMap(fn func(string) string, data map[string]interface{}, nonHMACDataKeys []string) (map[string]interface{}, error) {
+	if data == nil {
+		return nil, nil
+	}
+
+	newData := make(map[string]interface{}, len(data))
+	for k, v := range data {
+		if o, ok := v.(logical.OptMarshaler); ok {
+			marshaled, err := o.MarshalJSONWithOptions(&logical.MarshalOptions{
+				ValueHasher: fn,
+			})
+			if err != nil {
+				return nil, err
+			}
+			newData[k] = json.RawMessage(marshaled)
+		} else if strutil.StrListContains(nonHMACDataKeys, k) {
+			newData[k] = v
+		} else if t, ok := v.(time.Time); ok {
+			newData[k] = t.Format(time.RFC3339Nano)
```

Comment:
You don't need to call this specifically as marshaling JSON will automatically format time.Time as RFC3339Nano.

It might be better to do this in a type switch where the StrListContains case is moved down into default with what else is in the `else` block. Either way works though.